src/ddmd/root/filename.d

```diff
@@ -721,3 +734,142 @@ nothrow:
+        import core.stdc.errno: errno, EEXIST, ENOENT;
+
+        SetLastError(NO_ERROR);
+        const createRet = path.extendedPathThen!(p => CreateDirectoryW(&p[0],
+                                                                       null /*securityAttributes*/));
+        const lastError = GetLastError();
+
+        // Preserve compatibility with mkdir since the calling code expects
+        // errno to be set.
+        if (createRet == 0)
+        {
+            switch (lastError)
+            {
+                case ERROR_ALREADY_EXISTS:
+                    errno(EEXIST);
+                    break;
+                case ERROR_PATH_NOT_FOUND:
+                    errno(ENOENT);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts a path to one suitable to be passed to Win32 API
+    // functions that can deal with paths longer than 248
+    // characters then calls the supplied function on it.
+    // For more information:
+    // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx
+    package auto extendedPathThen(alias F)(const(char*) path)
+    {
+        import core.sys.windows.winbase: GetFullPathNameW;
+
+        wchar[1024] wpathBuf;
+        // Since the unicode Win32 APIs use UTF16, we need to convert
+        // from UTF8 to UTF16 first.
+        const wpath = path.toWStringz(wpathBuf);
+
+        // GetFullPathNameW expects a sized buffer to store the result in. Since we don't
+        // know how larget it has to be, we pass in null and get the needed buffer length
+        // as the return code.
+        const pathLength = GetFullPathNameW(&wpath[0],
+                                            0 /*length8*/,
+                                            null /*output buffer*/,
+                                            null /*filePartBuffer*/);
+        const wchar[] emptyPath;
+        if (pathLength == 0)
+        {
+            return F(emptyPath);
+        }
+
+        // wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
+
+        // +1 for the null terminator
+        const bufferLength = pathLength + prefix.length + 1;
+
+        wchar[1024] absBuf;
+        auto absPath = bufferLength > absBuf.length ? new wchar[bufferLength] : absBuf[];
+
+        absPath[0 .. prefix.length] = prefix[];
+
+        const absPathRet = GetFullPathNameW(&wpath[0],
+                                            absPath.length - prefix.length,
+                                            &absPath[prefix.length],
+                                            null /*filePartBuffer*/);
+
+        if (absPathRet == 0 || absPathRet > absPath.length)
```

Comment:
`absPathRet > absPath.length` can't really happen unless there's a bug, but also isn't the correct error check because you didn't pass a buffer of length `absPath.length`, you passed `absPath.length - prefix.length`

Reply:
Do you want me to change it to an assert?